doc/man/man3/DwmPingDestination.hh.3
.TH "DwmPingDestination.hh" 3 "libDwm-0.0.20240716" \" -*- nroff -*-
.ad l
.nh
.SH NAME
DwmPingDestination.hh \- \fBDwm::PingDestination\fP class definition\&.  

.SH SYNOPSIS
.br
.PP
\fR#include <string>\fP
.br
\fR#include 'DwmIpv4Address\&.hh'\fP
.br
\fR#include 'DwmSocket\&.hh'\fP
.br

.SS "Classes"

.in +1c
.ti -1c
.RI "class \fBDwm::PingDestination\fP"
.br
.RI "Common interface for an IPv4 destination we wish to ping with ICMP or TCP\&. "
.in -1c
.SH "Detailed Description"
.PP 
\fBDwm::PingDestination\fP class definition\&. 


.SH "Author"
.PP 
Generated automatically by Doxygen for libDwm-0\&.0\&.20240716 from the source code\&.


ChangeLog
2024-07-25  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as libDwm-0.9.45

	* README.md: mention libfmt dependency for the case where
	std::format is not available.

	* tagged as libDwm-0.9.44

	* classes/include/DwmGZIOCapable.hh: Documentation updates.
	* classes/include/DwmBZ2IOCapable.hh: Documentation updates.
	* classes/include/DwmASIOCapable.hh: Documentation updates.
	* classes/include/DwmDescriptorIOCapable.hh: Documentation updates.
	* classes/include/DwmFileIOCapable.hh: Documentation updates.
	* classes/include/DwmStreamIOCapable.hh: Documentation updates.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmDocMainPage.hh, README.md,
	doc/examples/DwmIOExample3.cc: Documentation updates.

2024-07-24  dwm  <dwm@nomnom.mcplex.net>

	* tagged as libDwm-0.9.43

	* mkrelease.sh: 'push --tags' should be 'push --follow-tags'

	* tagged as libDwm-0.9.42

	* classes/include/DwmSysLogger.hh: Fix pasto on platforms where
	we need libfmt (std:: should be fmt::), namely Raspberry Pi OS.

2024-07-24  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.41

	* classes/tests/TestThreadQueue.cc: Use max concurrency on
	multicore systems when running performance tests, just to
	add to the stress.  Also use C++14 digit separator (') in
	some numeric literals for readability.

2024-07-24  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/src/DwmSysLogger.cc, classes/include/DwmSysLogger.hh:
	Allow use of std::source_location, and use them in the Syslog
	and FSyslog macros.

	* Removed the following files since I don't need them here:

	classes/include/DwmHostPinger.hh
	classes/include/DwmIpv4IcmpDestinationUnreachable.hh
	classes/include/DwmIpv4IcmpEchoReply.hh
	classes/include/DwmIpv4IcmpEchoRequest.hh
	classes/include/DwmIpv4IcmpMessage.hh
	classes/include/DwmIpv4IcmpRedirect.hh
	classes/include/DwmIpv4IcmpSourceQuench.hh
	classes/include/DwmIpv4IcmpTimeExceeded.hh
	classes/include/DwmIpv4PacketHeader.hh
	classes/include/DwmIpv4TcpHeader.hh
	classes/include/DwmIpv4TcpPacket.hh
	classes/include/DwmIpv4TcpPayload.hh
	classes/include/DwmIpv4UdpHeader.hh
	classes/include/DwmIpv4UdpPacket.hh
	classes/include/DwmIpv4UdpPayload.hh
	classes/include/DwmPatricia.hh
	classes/include/DwmPingDestination.hh
	classes/include/DwmPingDestinationEntry.hh
	classes/include/DwmSocket.hh
	classes/include/DwmXmlAttribute.hh
	classes/include/DwmXmlElement.hh
	classes/src/DwmHostPinger.cc
	classes/src/DwmIpv4IcmpDestinationUnreachable.cc
	classes/src/DwmIpv4IcmpEchoReply.cc
	classes/src/DwmIpv4IcmpEchoRequest.cc
	classes/src/DwmIpv4IcmpMessage.cc
	classes/src/DwmIpv4IcmpRedirect.cc
	classes/src/DwmIpv4IcmpSourceQuench.cc
	classes/src/DwmIpv4IcmpTimeExceeded.cc
	classes/src/DwmIpv4PacketHeader.cc
	classes/src/DwmIpv4TcpHeader.cc
	classes/src/DwmIpv4TcpPacket.cc
	classes/src/DwmIpv4TcpPayload.cc
	classes/src/DwmIpv4UdpHeader.cc
	classes/src/DwmIpv4UdpPacket.cc
	classes/src/DwmIpv4UdpPayload.cc
	classes/src/DwmPingDestination.cc
	classes/src/DwmPingDestinationEntry.cc
	classes/src/DwmSocket.cc
	classes/src/DwmXmlAttribute.cc
	classes/src/DwmXmlElement.cc
	classes/src/Makefile
	classes/tests/Makefile
	classes/tests/TestDwmSocket.cc
	classes/tests/TestHostPinger.cc
	classes/tests/TestIpv4IcmpDestinationUnreachable.cc
	classes/tests/TestIpv4IcmpEchoReply.cc
	classes/tests/TestIpv4IcmpEchoRequest.cc
	classes/tests/TestIpv4IcmpRedirect.cc
	classes/tests/TestIpv4IcmpSourceQuench.cc
	classes/tests/TestIpv4IcmpTimeExceeded.cc
	classes/tests/TestIpv4PacketHeader.cc
	classes/tests/TestIpv4TcpHeader.cc
	classes/tests/TestIpv4TcpPacket.cc
	classes/tests/TestIpv4UdpHeader.cc
	classes/tests/TestIpv4UdpPacket.cc
	classes/tests/TestPatricia.cc
	classes/tests/ip.pcap
	classes/tests/tcp.pcap
	classes/tests/udp.pcap

2024-07-23  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* getvers.sh: bump shared library version to reflect ABI changes
	(vtables changed from removal of inheritance).

	* README.md: Document Read and Write requirements for
	ASIO with boost::asio::generic::stream_protocol::socket

	* classes/include/DwmIpv4IcmpMessage.hh,
	  classes/include/DwmIpv4PacketHeader.hh,
	  classes/include/DwmIpv4TcpHeader.hh,
	  classes/include/DwmIpv4TcpPacket.hh,
	  classes/include/DwmIpv4TcpPayload.hh,
	  classes/include/DwmIpv4UdpHeader.hh,
	  classes/include/DwmIpv4UdpPacket.hh,
	  classes/include/DwmIpv4UdpPayload.hh:
	No longer need to inherit from *Readable and *Writable classes.

	* classes/include/DwmASIOCapable.hh,
	  classes/include/DwmBZ2IOCapable.hh,
	  classes/include/DwmDescriptorIOCapable.hh,
	  classes/include/DwmFileIOCapable.hh,
	  classes/include/DwmGZIOCapable.hh,
	  classes/include/DwmStreamIOCapable.hh:
	Minor cleanup to concepts.

	* tagged as libDwm-0.9.40

	* doc/.gitignore: ignore libDwm.tag (generated by doxygen)

	* README.md: Add mkfbsdmnfst and mkdebcontrol as build
	dependencies.

2024-07-22  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Removed IN-PROGRESS file.

	* tagged as libDwm-0.9.39

	* Added README.md for the github mirror.

2024-07-21  Daniel McRobb  <dwm@thrip.mcplex.net>

	* tagged as libDwm-0.9.38

2024-07-21  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* doc/footer.html: Tweak to point at my web site.

	* classes/src/DwmMacAddress.cc: Use sizeof() where appropriate.

	* Remove doxygen-generated documentation from repository; it's
	build output that doesn't need to be in the repository.

2024-07-20  Daniel McRobb  <dwm@toto.mcplex.net>

	* tagged as libDwm-0.9.37

2024-07-20  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* classes/src/DwmProcessTableLex.lex,
	classes/src/DwmProcessTableParse.y: 'ps' output on FreeBSD shows
	a single hyphen ('-') instead of a ctime() string for lstart for
	some processes.  Patch this by substituting the lstart from the
	nearest ancestor of the process that has a valid lstart.

	* classes/include/DwmBootTime.hh, classes/include/DwmUptime.hh
	classes/src/DwmBootTime.cc, classes/src/DwmUptime.cc:

classes/src/DwmPingDestination.cc
//===========================================================================
// @(#) $DwmPath$
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2008
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file DwmPingDestination.cc
//!  \author Daniel W. McRobb
//!  \brief Dwm::PingDestination class implementation
//---------------------------------------------------------------------------

extern "C" {
  #include <sys/types.h>
  #include <netinet/in_systm.h>
  #include <netinet/in.h>
  #include <netinet/ip.h>
  #include <errno.h>
  #include <netdb.h>
  #include <string.h>  // for memset()
}

#include <sstream>
#include <string>

#include "DwmIpv4PacketHeader.hh"
#include "DwmIpv4IcmpEchoRequest.hh"
#include "DwmIpv4TcpHeader.hh"
#include "DwmStringUtils.hh"
#include "DwmTimeValue.hh"
#include "DwmPingDestination.hh"

using namespace std;

namespace Dwm {

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  PingDestination::PingDestination()
      : _addr(), _port(0), _primed(false), _sequenceNumber(0)
  {}

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  PingDestination::PingDestination(const string & destination)
      : _addr(), _port(0), _primed(false), _sequenceNumber(0)
  {
    istringstream  is(destination);
    string         s;
    if (getline(is, s, ':')) {
      _addr = s;
      if (getline(is, s)) {
        uint16_t  port;
        if (StringUtils::StringTo(s, port)) {
          _port = port;
        }
      }
    }
    if (_addr == Ipv4Address("255.255.255.255")) {
      is.clear();
      is.seekg(0, ios::beg);
      string  hostName;
      if (getline(is, hostName, ':')) {
        struct hostent  *hostEntry = gethostbyname(hostName.c_str());
        if (hostEntry) {
          Address(*(ipv4addr_t *)hostEntry->h_addr_list[0]);
          _hostname = hostName;
        }
        else {
          cerr << "host '" << hostName << "' not found" << endl;
        }
      }
    }
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  PingDestination::PingDestination(const Ipv4Address & addr, uint16_t port)
      : _addr(addr), _port(port), _primed(false), _sequenceNumber(0)
  {}
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  const Ipv4Address & PingDestination::Address() const
  {
    return(_addr);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  const Ipv4Address & PingDestination::Address(const Ipv4Address & address)
  {
    _addr = address;
    return(_addr);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  uint16_t PingDestination::Port() const
  {
    return(_port);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  const std::string & PingDestination::Hostname() const
  {
    return(_hostname);
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  const std::string & PingDestination::Hostname(const std::string & hostname)
  {
    _hostname = hostname;
    return(_hostname);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool PingDestination::SendTo(Socket & socket, const Ipv4Address & myAddress,
                               uint16_t identifier) const
  {
    string  packet;
    GetPacket(packet, myAddress, identifier);
    struct sockaddr_in  destination;
    memset(&destination, 0, sizeof(destination));
    destination.sin_family = PF_INET;
#ifndef __linux__
    destination.sin_len = sizeof(struct sockaddr_in);
#endif
    destination.sin_addr.s_addr = _addr.Raw();

    return(socket.SendTo(packet.c_str(), packet.length(), 0, destination));
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool PingDestination::operator < (const PingDestination & pd) const
  {
    return((string)(*this) < (string)(pd));
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  bool PingDestination::operator == (const PingDestination & pd) const
  {
    return((string)(*this) == (string)(pd));
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  PingDestination::operator string () const
  {
    ostringstream  os;
    os << *this;
    return(os.str());
  }

  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  std::ostream & operator << (std::ostream & os, const PingDestination & pd)
  {
    if (os) {
      os << pd._addr;
      if (pd._port) {
        os << ':' << pd._port;
      }
    }
    return(os);
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  void PingDestination::GetPacket(std::string & packet,
                                  const Ipv4Address & myAddress,
                                  uint16_t identifier) const
  {
    if (_port) {
      GetTcpPacket(packet, myAddress, identifier);
    }
    else {
      GetIcmpPacket(packet, myAddress, identifier);
    }
    return;
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  void PingDestination::GetIcmpPacket(std::string & packet,
                                      const Ipv4Address & myAddress,
                                      uint16_t identifier) const
  {
    Ipv4PacketHeader  packetHeader;
    packetHeader.HeaderLength(20);
    packetHeader.TypeOfService(0);
    packetHeader.Id(identifier);
    packetHeader.FragmentOffset(0);
    packetHeader.TimeToLive(250);
    packetHeader.Protocol(1);
    packetHeader.SourceAddress(myAddress);
    packetHeader.DestinationAddress(_addr);

    Ipv4IcmpEchoRequest  echoRequest;
    echoRequest.Identifier(identifier);
    echoRequest.SequenceNumber(++_sequenceNumber);
    if (! _sequenceNumber) {
      ++_sequenceNumber;  // handle rollover; we only want to send a 0 once.
    }
    echoRequest.SetData(TimeValue(true));
    echoRequest.SetChecksum();
      
    packetHeader.TotalLength(packetHeader.StreamedLength() + 
                             echoRequest.StreamedLength());
    packetHeader.SetChecksum();

    ostringstream  os;
    packetHeader.Write(os);
    echoRequest.Write(os);
    packet = os.str();
    
    return;
  }
  
  //--------------------------------------------------------------------------
  //!  
  //--------------------------------------------------------------------------
  void PingDestination::GetTcpPacket(std::string & packet, 
                                     const Ipv4Address & myAddress, 
                                     uint16_t identifier) const
  {
    Ipv4PacketHeader  packetHeader;
    packetHeader.HeaderLength(20);
    packetHeader.TypeOfService(0);
    packetHeader.TotalLength(40);
    packetHeader.Id(identifier);
    packetHeader.FragmentOffset(0);
    packetHeader.TimeToLive(250);
    packetHeader.Protocol(6);
    packetHeader.SourceAddress(myAddress);
    packetHeader.DestinationAddress(_addr);
    packetHeader.SetChecksum();

    Ipv4TcpHeader  tcpHeader;
    tcpHeader.SourcePort(1);
    tcpHeader.DestinationPort(_port);
    if (_sequenceNumber == 0xFFFF) {
      _sequenceNumber = 1;  // handle rollover.  We only want to send 0 once.
    }
    tcpHeader.SequenceNumber(_sequenceNumber);
    tcpHeader.AckNumber(_sequenceNumber);
    ++_sequenceNumber;
    tcpHeader.DataOffset(20);
    if (! _primed) {
      tcpHeader.Flags(TH_SYN);
      _primed = true;
    }
    else {
      tcpHeader.Flags(TH_ACK);
    }
    tcpHeader.Window(1024);
    tcpHeader.SetChecksum(packetHeader, "");

    ostringstream  os;
    packetHeader.Write(os);
    tcpHeader.Write(os);
    packet = os.str();
    return;
  }
  
}  // namespace Dwm


//---------------------------- emacs settings -----------------------------
//  Local Variables:
//  mode: C++
//  tab-width: 2
//  indent-tabs-mode: nil
//  c-basic-offset: 2
//  End:
//-------------------------------------------------------------------------
